src/connection.sml

```diff
@@ -1,68 +1,194 @@
-                val mP =  PolyML.makestring messageType
-                val lP =  PolyML.makestring length
-            in 
-                case messageType of
-                    "R" => print "Password"
-                  | "Z" => print "ReadyForQuery"
-                  | "E" => print "Error"
-                  | _ => print "NAY";
-                Socket.close socket
-            end
+  fun bytesToInt16 (bytes: Word8Vector.vector) =
+    let
+      open Word8
+      open Word8VectorSlice
+      val slices = full bytes
+      val first = sub (slices, 0)
+      val second = sub (slices, 1)
+    in
+      Word8.toInt (orb (second, (<<(first, 0wx8))))
+    end
+
+  fun startup (socket: Socket.active INetSock.stream_sock) =
+    let
+      val message = convertString "user\000admin\000database\000socket\000\000"
+      val length = intToBytes 36
+      (* Word8Vector.fromList [ 0wx50, 0wx0, 0wx0, 0wx0 ] *)
+      val protocolVersion = Word8Vector.fromList [0w0, 0w3, 0w0, 0w0]
+      val buffer = Word8VectorSlice.full
+        (Word8Vector.concat [length, protocolVersion, message])
+    (* val buf = PolyML.makestring buffer *)
+    in
+      (* print buf; *)
+      (* print "\n\n"; *)
+
+      Socket.sendVec (socket, buffer)
+    end
+
+  fun password (message: Word8Vector.vector) =
+
+    let
+      val (vec, _, _) = Word8VectorSlice.base
+        (Word8VectorSlice.slice (message, 0, SOME 4))
+    in
+      case (bytesToInt vec) of
+        0 => OK
+      | 5 => ERROR "MD5 method not implemented."
+      | _ => ERROR "Unindentified password method."
+    end
+
+  (* let 
+      val inner = 
+          MD5.final (MD5.update(MD5.init, convertString "admin\000admin"))
+      val pwd =
+          Word8Vector.concat
+              [(convertString "md5"),
+               (MD5.final (MD5.update (MD5.init, (Word8Vector.concat [ inner, convertString salt ]))))]
+      val serialized = PolyML.makestring pwd
+  in 
+      print serialized
+  end *)
+
+  fun execute (socket: Socket.active INetSock.stream_sock) query =
+    let
+      val queryFlag = Word8VectorSlice.full
+        (Word8Vector.fromList [Byte.charToByte #"Q"])
+      val serializedQuery = Word8VectorSlice.full (convertString query)
+      val length = intToBytes (4 + 1 + Word8VectorSlice.length serializedQuery)
+      val message = Word8VectorSlice.full (Word8VectorSlice.concat
+        [ queryFlag
+        , Word8VectorSlice.full length
+        , serializedQuery
+        , Word8VectorSlice.full (convertString "\000")
+        ])
+      val str = PolyML.makestring message
+    in
+      print str;
+      Socket.sendVec (socket, message);
+      OK
+    end
+
+  fun parser (socket: Socket.active INetSock.stream_sock) =
+    let
+      val messageType = (* Byte.bytesToString ( *) Socket.recvVec (socket, 1)
+      val length = (* bytesToInt( *) Socket.recvVec (socket, 4)
+      val mP =
+        "Type: " ^ PolyML.makestring messageType ^ " -> "
```

Comment:
this hard codes your code to Poly. This is dangerous for a PG library

Reply:
Yes, this is purely for debugging